src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,183 @@
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class ErrorListTableDataSource : ITableDataSource
+    {
+        private readonly object _lockObj = new object();
+        private readonly IServiceProvider _serviceProvider;
+
+        private TableSubscription _tableSubscription;
+
+        public string SourceTypeIdentifier => StandardTableDataSources.ErrorTableDataSource;
+
+        public string Identifier => "NuGetRestoreManagerListTable";
+
+        public string DisplayName => "NuGet_Restore_Manager_Table_Data_Source";
+
+        public IErrorList _errorList;
+        public ITableManager _tableManager;
+        private bool _initialized;
+
+        [ImportingConstructor]
+        public ErrorListTableDataSource(
+            [Import(typeof(SVsServiceProvider))]
+            IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            _serviceProvider = serviceProvider;
+        }
+
+        public IDisposable Subscribe(ITableDataSink sink)
+        {
+            lock (_lockObj)
+            {
+                _tableSubscription = new TableSubscription(sink, _lockObj);
+                return _tableSubscription;
+            }
+        }
+
+        /// <summary>
+        /// Clear only nuget entries.
+        /// </summary>
+        public void ClearNuGetEntries()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            lock (_lockObj)
+            {
+                var sink = _tableSubscription?.TableDataSink;
+
+                if (sink != null)
+                {
+                    var entries = _errorList.TableControl?.Entries?.Where(IsNuGetEntry).ToArray()
+                        ?? new ITableEntryHandle[0];
+
+                    sink.RemoveEntries(entries);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add error list entries
+        /// </summary>
+        public void AddEntries(params ErrorListTableEntry[] entries)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            lock (_lockObj)
+            {
+                var sink = _tableSubscription?.TableDataSink;
+
+                if (sink != null)
+                {
+                    sink.AddEntries(entries.ToList(), removeAllEntries: false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Show error window.
+        /// </summary>
+        public void BringToFront()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            // Give the error list focus.
+            var vsErrorList = _errorList as IVsErrorList;
+            vsErrorList?.BringToFront();
+        }
+
+        private void EnsureInitialized()
+        {
+            if (!_initialized)
```

Comment:
good catch, looks like this could end up subscribing twice, I'll make sure it is checked within the lock